File: ios/Presentation/Profile/ProfileViewModel.swift
import Foundation
import SwiftUI
import Combine

enum ProfileState: Equatable { case idle, loading, loaded, error(String) }
enum ProfileTab: String, CaseIterable { case recipes, logs, saved }
enum VisibilityFilter: String, CaseIterable {
    case all, publicOnly, privateOnly

    var title: String {
        switch self {
        case .all: return String(localized: "filter.all")
        case .publicOnly: return String(localized: "filter.public")
        case .privateOnly: return String(localized: "filter.private")
        }
    }
}

enum SavedContentFilter: String, CaseIterable {
    case all, recipes, logs

    var title: String {
        switch self {
        case .all: return String(localized: "filter.all")
        case .recipes: return String(localized: "profile.recipes")
        case .logs: return String(localized: "profile.logs")
        }
    }
}

@MainActor
final class ProfileViewModel: ObservableObject {
    @Published private(set) var state: ProfileState = .idle
    @Published private(set) var profile: UserProfile?
    @Published private(set) var myProfile: MyProfile?
    @Published private(set) var isOwnProfile: Bool
    @Published private(set) var recipes: [RecipeSummary] = []
    @Published private(set) var logs: [FeedLogItem] = []
    @Published private(set) var savedRecipes: [RecipeSummary] = []
    @Published private(set) var savedLogs: [FeedLogItem] = []
    @Published private(set) var isLoadingContent = false
    @Published private(set) var hasMoreRecipes = true
    @Published private(set) var hasMoreLogs = true
    @Published private(set) var hasMoreSavedRecipes = true
    @Published private(set) var hasMoreSavedLogs = true
    @Published private(set) var savedContentVersion: Int = 0
    @Published var selectedTab: ProfileTab = .recipes
    @Published var visibilityFilter: VisibilityFilter = .all
    @Published var savedContentFilter: SavedContentFilter = .all

    private let userId: String?
    private let userRepository: UserRepositoryProtocol
    private let logRepository: CookingLogRepositoryProtocol
    private let savedContentRepository: SavedContentRepositoryProtocol
    private var recipesNextCursor: String?
    private var logsNextCursor: String?
    private var savedRecipesNextCursor: String?
    private var savedLogsNextCursor: String?
    private var cancellables = Set<AnyCancellable>()

    init(
        userId: String? = nil,
        userRepository: UserRepositoryProtocol = UserRepository(),
        logRepository: CookingLogRepositoryProtocol = CookingLogRepository(),
        savedContentRepository: SavedContentRepositoryProtocol = SavedContentRepository()
    ) {
        self.userId = userId
        self.userRepository = userRepository
        self.logRepository = logRepository
        self.savedContentRepository = savedContentRepository
        self.isOwnProfile = userId == nil
        setupSaveStateObserver()
    }

    private func setupSaveStateObserver() {
        guard isOwnProfile else {
            #if DEBUG
            print("[ProfileVM] NOT own profile - skipping observer setup")
            #endif
            return
        }

        #if DEBUG
        print("========== [ProfileVM] SETTING UP OBSERVERS ==========")
        #endif

        // Observe SavedItemsManager for saved recipes changes
        SavedItemsManager.shared.$savedRecipes
            .receive(on: DispatchQueue.main)
            .sink { [weak self] recipes in
                guard let self = self else {
                    #if DEBUG
                    print("[ProfileVM] RECIPE OBSERVER - self is nil!")
                    #endif
                    return
                }
                #if DEBUG
                print("========== [ProfileVM] RECIPE OBSERVER FIRED ==========")
                print("[ProfileVM] Received \(recipes.count) recipes from SavedItemsManager")
                for (index, r) in recipes.prefix(5).enumerated() {
                    print("[ProfileVM]   [\(index)] \(r.id): \(r.coverImageUrl ?? "NIL")")
                }
                print("[ProfileVM] BEFORE assignment - self.savedRecipes.count: \(self.savedRecipes.count)")
                #endif
                self.savedRecipes = recipes
                self.savedContentVersion += 1
                #if DEBUG
                print("[ProfileVM] AFTER assignment - self.savedRecipes.count: \(self.savedRecipes.count)")
                print("[ProfileVM] savedContentVersion: \(self.savedContentVersion)")
                print("========== [ProfileVM] RECIPE OBSERVER END ==========")
                #endif
            }
            .store(in: &cancellables)

        // Observe SavedItemsManager for saved logs changes
        SavedItemsManager.shared.$savedLogs
            .receive(on: DispatchQueue.main)
            .sink { [weak self] logs in
                guard let self = self else {
                    #if DEBUG
                    print("[ProfileVM] LOG OBSERVER - self is nil!")
                    #endif
                    return
                }
                #if DEBUG
                print("========== [ProfileVM] LOG OBSERVER FIRED ==========")
                print("[ProfileVM] Received \(logs.count) logs from SavedItemsManager")
                for (index, log) in logs.prefix(5).enumerated() {
                    print("[ProfileVM]   [\(index)] \(log.id): \(log.thumbnailUrl ?? "NIL")")
                }
                print("[ProfileVM] BEFORE assignment - self.savedLogs.count: \(self.savedLogs.count)")
                #endif
                self.savedLogs = logs
                self.savedContentVersion += 1
                #if DEBUG
                print("[ProfileVM] AFTER assignment - self.savedLogs.count: \(self.savedLogs.count)")
                print("[ProfileVM] savedContentVersion: \(self.savedContentVersion)")
                print("========== [ProfileVM] LOG OBSERVER END ==========")
                #endif
            }
            .store(in: &cancellables)

        #if DEBUG
        print("[ProfileVM] Observers setup complete")
        #endif
    }

    func loadProfile() {
        state = .loading
        Task {
            if isOwnProfile { await loadMyProfile() }
            else if let id = userId { await loadUserProfile(id) }
        }
    }

    func loadContent() {
        #if DEBUG
        print("[Profile] loadContent called for tab: \(selectedTab)")
        #endif
        Task {
            switch selectedTab {
            case .recipes: await loadRecipes(refresh: true)
            case .logs: await loadLogs(refresh: true)
            case .saved: await loadSavedContent(refresh: true)
            }
        }
    }

    var savedCount: Int {
        // Use local published properties for proper SwiftUI updates
        savedRecipes.count + savedLogs.count
    }

    func toggleFollow() async {
        guard let profile = profile, !isOwnProfile else { return }
        let wasFollowing = profile.isFollowing
        updateFollowState(!wasFollowing)
        let result = wasFollowing ? await userRepository.unfollow(userId: profile.id) : await userRepository.follow(userId: profile.id)
        if case .failure = result { updateFollowState(wasFollowing) }
    }

    func blockUser() async {
        guard let profile = profile else { return }
        let result = await userRepository.blockUser(userId: profile.id)
        if case .success = result {
            self.profile = UserProfile(
                id: profile.id,
                username: profile.username,
                displayName: profile.displayName,
                avatarUrl: profile.avatarUrl,
                bio: profile.bio,
                level: profile.level,
                levelName: profile.levelName,
                recipeCount: profile.recipeCount,
                logCount: profile.logCount,
                followerCount: profile.followerCount,
                followingCount: profile.followingCount,
                youtubeUrl: profile.youtubeUrl,
                instagramHandle: profile.instagramHandle,
                isFollowing: false,
                isFollowedBy: profile.isFollowedBy,
                isBlocked: true,
                createdAt: profile.createdAt
            )
        }
    }

    func reportUser(reason: ReportReason) async {
        guard let profile = profile else { return }
        let result = await userRepository.reportUser(userId: profile.id, reason: reason)
        #if DEBUG
        if case .success = result {
            print("[Profile] Reported user \(profile.id) for: \(reason.rawValue)")
        }
        #endif
    }

    private func loadMyProfile() async {
        let result = await userRepository.getMyProfile()
        switch result {
        case .success(let profile):
            myProfile = profile
            state = .loaded
            loadContent()
        case .failure(let error):
            state = .error(error.localizedDescription)
        }
    }

    private func loadUserProfile(_ id: String) async {
        let result = await userRepository.getUserProfile(id: id)
        switch result {
        case .success(let profile): self.profile = profile; state = .loaded; loadContent()
        case .failure(let error): state = .error(error.localizedDescription)
        }
    }

    private func loadRecipes(refresh: Bool) async {
        guard let id = userId ?? myProfile?.id else { return }
        isLoadingContent = true
        defer { isLoadingContent = false }
        let result = await userRepository.getUserRecipes(userId: id, cursor: refresh ? nil : recipesNextCursor)
        if case .success(let response) = result {
            recipes = refresh ? response.content : recipes + response.content
            recipesNextCursor = response.nextCursor
            hasMoreRecipes = response.hasMore
        }
    }

    private func loadLogs(refresh: Bool) async {
        guard let id = userId ?? myProfile?.id else { return }
        isLoadingContent = true
        defer { isLoadingContent = false }
        let result = await logRepository.getUserLogs(userId: id, cursor: refresh ? nil : logsNextCursor)
        if case .success(let response) = result {
            logs = refresh ? response.content : logs + response.content
            logsNextCursor = response.nextCursor
            hasMoreLogs = response.hasMore
        }
    }

    private func updateFollowState(_ isFollowing: Bool) {
        guard let p = profile else { return }
        profile = UserProfile(
            id: p.id,
            username: p.username,
            displayName: p.displayName,
            avatarUrl: p.avatarUrl,
            bio: p.bio,
            level: p.level,
            levelName: p.levelName,
            recipeCount: p.recipeCount,
            logCount: p.logCount,
            followerCount: p.followerCount + (isFollowing ? 1 : -1),
            followingCount: p.followingCount,
            youtubeUrl: p.youtubeUrl,
            instagramHandle: p.instagramHandle,
            isFollowing: isFollowing,
            isFollowedBy: p.isFollowedBy,
            isBlocked: p.isBlocked,
            createdAt: p.createdAt
        )
    }

    private func loadSavedContent(refresh: Bool) async {
        guard isOwnProfile else { return }
        isLoadingContent = true
        defer { isLoadingContent = false }

        #if DEBUG
        print("[Profile] Loading saved content, refresh=\(refresh)")
        #endif

        // Delegate to SavedItemsManager which is the single source of truth
        await SavedItemsManager.shared.fetchAllSavedContent()

        #if DEBUG
        print("[Profile] Saved content loaded via SavedItemsManager")
        print("[Profile]   - Recipes: \(SavedItemsManager.shared.savedRecipes.count)")
        print("[Profile]   - Logs: \(SavedItemsManager.shared.savedLogs.count)")
        #endif
    }

    func loadMoreContent() {
        Task {
            switch selectedTab {
            case .recipes: await loadRecipes(refresh: false)
            case .logs: await loadLogs(refresh: false)
            case .saved: await loadSavedContent(refresh: false)
            }
        }
    }

    /// Refresh saved content (called when saved tab appears)
    func refreshSavedContentIfNeeded() async {
        guard isOwnProfile, !isLoadingContent else { return }

        #if DEBUG
        print("[Profile] refreshSavedContentIfNeeded: forcing refresh")
        #endif

        // Always refresh to ensure we have the latest saved content
        await loadSavedContent(refresh: true)
    }

    /// Reset all state (called when user changes)
    func reset() {
        #if DEBUG
        print("[Profile] Resetting profile state")
        #endif
        state = .idle
        profile = nil
        myProfile = nil
        recipes = []
        logs = []
        savedRecipes = []
        savedLogs = []
        savedContentVersion = 0
        recipesNextCursor = nil
        logsNextCursor = nil
        savedRecipesNextCursor = nil
        savedLogsNextCursor = nil
        hasMoreRecipes = true
        hasMoreLogs = true
        hasMoreSavedRecipes = true
        hasMoreSavedLogs = true
        selectedTab = .recipes
    }
}

File: android/app/src/main/java/com/cookstemma/app/ui/screens/profile/ProfileViewModel.kt
package com.cookstemma.app.ui.screens.profile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.cookstemma.app.data.repository.UserRepository
import com.cookstemma.app.domain.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

enum class ProfileTab { RECIPES, LOGS, SAVED }

enum class VisibilityFilter(val title: String) {
    ALL("All"),
    PUBLIC("Public"),
    PRIVATE("Private")
}

enum class SavedContentFilter(val title: String) {
    ALL("All"),
    RECIPES("Recipes"),
    LOGS("Logs")
}

data class ProfileUiState(
    val isLoading: Boolean = false,
    val isLoadingContent: Boolean = false,
    val error: String? = null,
    val userId: String? = null,
    val avatarUrl: String? = null,
    val username: String = "",
    val displayName: String? = null,
    val bio: String? = null,
    val level: Int? = null,
    val levelName: String? = null,
    val levelProgress: Double? = null,
    val recipeCount: Int = 0,
    val logCount: Int = 0,
    val savedCount: Int = 0,
    val followerCount: Int = 0,
    val followingCount: Int = 0,
    val isFollowing: Boolean = false,
    val isBlocked: Boolean = false,
    val socialLinks: SocialLinks? = null,
    val youtubeUrl: String? = null,
    val instagramHandle: String? = null,
    val selectedTab: ProfileTab = ProfileTab.RECIPES,
    val visibilityFilter: VisibilityFilter = VisibilityFilter.ALL,
    val savedContentFilter: SavedContentFilter = SavedContentFilter.ALL,
    val recipes: List<RecipeSummary> = emptyList(),
    val logs: List<CookingLogSummary> = emptyList(),
    val savedRecipes: List<RecipeSummary> = emptyList(),
    val savedLogs: List<CookingLog> = emptyList(),
    val blockSuccess: Boolean = false,
    val reportSuccess: Boolean = false
) {
    val localizedLevelName: String get() = LevelName.displayName(levelName)
}

@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val userRepository: UserRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(ProfileUiState())
    val uiState: StateFlow<ProfileUiState> = _uiState.asStateFlow()

    private var currentUserId: String? = null
    private var isOwnProfile: Boolean = false

    fun loadProfile(userId: String?) {
        currentUserId = userId
        isOwnProfile = userId == null

        viewModelScope.launch {
            val profileFlow = if (isOwnProfile) userRepository.getMyProfile() else userRepository.getUserProfile(userId!!)
            profileFlow.collect { result ->
                when (result) {
                    is Result.Loading -> _uiState.update { it.copy(isLoading = true, error = null) }
                    is Result.Success -> {
                        when (val data = result.data) {
                            is MyProfile -> _uiState.update {
                                it.copy(
                                    isLoading = false,
                                    userId = data.id,
                                    avatarUrl = data.avatarUrl,
                                    username = data.username ?: "",
                                    displayName = data.displayName,
                                    bio = data.bio,
                                    level = data.level,
                                    levelName = data.levelName,
                                    levelProgress = data.levelProgress,
                                    recipeCount = data.recipeCount,
                                    logCount = data.logCount,
                                    savedCount = data.savedCount,
                                    followerCount = data.followerCount,
                                    followingCount = data.followingCount,
                                    socialLinks = data.socialLinks,
                                    youtubeUrl = data.youtubeUrl,
                                    instagramHandle = data.instagramHandle
                                )
                            }
                            is UserProfile -> _uiState.update {
                                it.copy(
                                    isLoading = false,
                                    userId = data.id,
                                    avatarUrl = data.avatarUrl,
                                    username = data.username ?: "",
                                    displayName = data.displayName,
                                    bio = data.bio,
                                    level = data.level,
                                    levelName = data.levelName,
                                    recipeCount = data.recipeCount,
                                    logCount = data.logCount,
                                    followerCount = data.followerCount,
                                    followingCount = data.followingCount,
                                    isFollowing = data.isFollowing,
                                    isBlocked = data.isBlocked,
                                    socialLinks = data.socialLinks,
                                    youtubeUrl = data.youtubeUrl,
                                    instagramHandle = data.instagramHandle
                                )
                            }
                        }
                        loadContent()
                    }
                    is Result.Error -> _uiState.update { it.copy(isLoading = false, error = result.exception.message) }
                }
            }
        }
    }

    fun selectTab(tab: ProfileTab) {
        _uiState.update { it.copy(selectedTab = tab) }
        loadContent()
    }

    fun toggleFollow() {
        val userId = currentUserId ?: return
        viewModelScope.launch {
            val wasFollowing = _uiState.value.isFollowing
            _uiState.update { it.copy(isFollowing = !wasFollowing, followerCount = it.followerCount + if (wasFollowing) -1 else 1) }
            val result = if (wasFollowing) userRepository.unfollowUser(userId) else userRepository.followUser(userId)
            if (result is Result.Error) {
                _uiState.update { it.copy(isFollowing = wasFollowing, followerCount = it.followerCount + if (wasFollowing) 1 else -1) }
            }
        }
    }

    fun setVisibilityFilter(filter: VisibilityFilter) {
        _uiState.update { it.copy(visibilityFilter = filter) }
        loadContent()
    }

    fun setSavedContentFilter(filter: SavedContentFilter) {
        _uiState.update { it.copy(savedContentFilter = filter) }
        loadSavedContent()
    }

    fun loadContent() {
        val userId = currentUserId ?: _uiState.value.userId ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingContent = true) }
            when (_uiState.value.selectedTab) {
                ProfileTab.RECIPES -> {
                    userRepository.getUserRecipes(userId, null).collect { result ->
                        if (result is Result.Success) {
                            _uiState.update { it.copy(recipes = result.data.content, isLoadingContent = false) }
                        } else {
                            _uiState.update { it.copy(isLoadingContent = false) }
                        }
                    }
                }
                ProfileTab.LOGS -> {
                    userRepository.getUserLogs(userId, null).collect { result ->
                        if (result is Result.Success) {
                            _uiState.update { it.copy(logs = result.data.content, isLoadingContent = false) }
                        } else {
                            _uiState.update { it.copy(isLoadingContent = false) }
                        }
                    }
                }
                ProfileTab.SAVED -> {
                    loadSavedContent()
                }
            }
        }
    }

    private fun loadSavedContent() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingContent = true) }
            // Load saved recipes
            userRepository.getSavedRecipes(null).collect { result ->
                if (result is Result.Success) {
                    _uiState.update { it.copy(savedRecipes = result.data.content) }
                }
            }
            // Load saved logs
            userRepository.getSavedLogs(null).collect { result ->
                if (result is Result.Success) {
                    _uiState.update { it.copy(savedLogs = result.data.content, isLoadingContent = false) }
                } else {
                    _uiState.update { it.copy(isLoadingContent = false) }
                }
            }
        }
    }

    fun blockUser() {
        val userId = currentUserId ?: return
        viewModelScope.launch {
            val wasBlocked = _uiState.value.isBlocked
            // Optimistic update
            _uiState.update { it.copy(isBlocked = !wasBlocked) }
            
            val result = if (wasBlocked) {
                userRepository.unblockUser(userId)
            } else {
                userRepository.blockUser(userId)
            }
            
            when (result) {
                is Result.Success -> {
                    if (!wasBlocked) {
                        _uiState.update { it.copy(blockSuccess = true) }
                    }
                }
                is Result.Error -> {
                    // Revert on failure
                    _uiState.update { it.copy(isBlocked = wasBlocked) }
                }
                else -> {}
            }
        }
    }

    fun reportUser(reason: String) {
        val userId = currentUserId ?: return
        viewModelScope.launch {
            val result = userRepository.reportUser(userId, reason)
            if (result is Result.Success) {
                _uiState.update { it.copy(reportSuccess = true) }
            }
        }
    }

    fun clearBlockSuccess() {
        _uiState.update { it.copy(blockSuccess = false) }
    }

    fun clearReportSuccess() {
        _uiState.update { it.copy(reportSuccess = false) }
    }
}
